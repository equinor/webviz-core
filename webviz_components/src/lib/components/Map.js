import React, {Component} from 'react';
import PropTypes from 'prop-types';
import FlowMap from '../private-components/map/flow_map';
import Map2D from '../private-components/map/map2d';

const getIndexies = layers => {
    const index = {};
    layers.forEach(kLayer => {
        kLayer.forEach(({k, i, j, ...layer}) => {
            if (!index[k]) {
                index[k] = {};
            }
            if (!index[k][i]) {
                index[k][i] = {};
            }
            if (!index[k][i][j]) {
                index[k][i][j] = {};
                index[k][i][j]['FLOWI+'] = layer['FLOWI+'];
                index[k][i][j]['FLOWJ+'] = layer['FLOWJ+'];
            }
        });
    });
    return index;
};

const addNegativeFlow = ({layers, indexies}) =>
    layers.map(kLayer =>
        kLayer.map(({i, j, k, ...layer}) => {
            let FLOWInegative = 0;
            let FLOWJnegative = 0;
            if (
                indexies[k][i - 1] &&
                indexies[k][i - 1][j] &&
                indexies[k][i - 1][j]['FLOWI+'] !== undefined
            ) {
                FLOWInegative = indexies[k][i - 1][j]['FLOWI+'];
            }
            if (
                indexies[k][i][j - 1] &&
                indexies[k][i][j - 1]['FLOWJ+'] !== undefined
            ) {
                FLOWJnegative = indexies[k][i][j - 1]['FLOWJ+'];
            }
            return {
                ...layer,
                k,
                i,
                j,
                'FLOWI-': FLOWInegative,
                'FLOWJ-': FLOWJnegative,
            };
        })
    );

export function makeFlowLayers(data) {
    const {i, j, k, x0, y0, x1, y1, x2, y2, x3, y3, value} = data;
    const FLOWIplus = data['FLOWI+'];
    const FLOWJplus = data['FLOWJ+'];
    const keys = Object.keys(data.i);
    const layers = [];

    keys.forEach(key => {
        const kValue = k[key];
        if (!layers[kValue]) {
            layers[kValue] = [];
        }
        layers[kValue].push({
            i: i[key],
            j: j[key],
            k: k[key],
            points: [
                [x0[key], y0[key]],
                [x1[key], y1[key]],
                [x2[key], y2[key]],
                [x3[key], y3[key]],
            ],
            value: value[key],
            'FLOWI+': FLOWIplus[key],
            'FLOWJ+': FLOWJplus[key],
        });
    });
    const indexies = getIndexies(layers);
    return addNegativeFlow({layers, indexies});
}

export const make2DLayers = ({
    i,
    j,
    k,
    x0,
    y0,
    x1,
    y1,
    x2,
    y2,
    x3,
    y3,
    value,
}) => {
    const layers = [];
    Object.keys(i).forEach(key => {
        const kValue = k[key];
        if (!layers[kValue]) {
            layers[kValue] = [];
        }
        layers[kValue].push({
            i: i[key],
            j: j[key],
            k: k[key],
            points: [
                [x0[key], y0[key]],
                [x1[key], y1[key]],
                [x2[key], y2[key]],
                [x3[key], y3[key]],
            ],
            value: value[key],
        });
    });
    return layers;
};

const initFlowMap = ({
    canvasSelector,
    elementSelector,
    data,
    height,
    layerNames,
}) => {
    const layers = makeFlowLayers(data);
    const map = new FlowMap({
        canvasSelector,
        elementSelector,
        layers,
        layerNames,
        height,
    });
    map.init();
};

const init2DMap = ({elementSelector, data, height, layerNames}) => {
    const layers = make2DLayers(data);
    const map = new Map2D({
        elementSelector,
        layers,
        layerNames,
        height,
    });
    map.init();
};

const parseData = data => (typeof data === 'string' ? JSON.parse(data) : data);

const shouldRenderFlowMap = data => Boolean(data['FLOWI+']);

class Map extends Component {
    constructor(props) {
        super(props);
        this.canvas = null;
        this.canvasId = `canvas-${props.id}`;
        this.elementId = `container-${props.id}`;
    }

    componentDidMount() {
        if (this.canvas) {
            const {data, height, layerNames} = this.props;
            const parsedData = parseData(data);
            const isFlowMap = shouldRenderFlowMap(parsedData);
            const canvasSelector = `#${this.canvasId}`;
            const elementSelector = `#${this.elementId}`;
            if (isFlowMap) {
                initFlowMap({
                    canvasSelector,
                    elementSelector,
                    data: parsedData,
                    height,
                    layerNames,
                });
            } else {
                init2DMap({
                    elementSelector,
                    data: parsedData,
                    height,
                    layerNames,
                });
            }
        }
    }

    render() {
        /* eslint-disable react/destructuring-assignment, react/prop-types */

        const {props} =
            this.props._dashprivate_layout &&
            this.props._dashprivate_layout.props
                ? this.props._dashprivate_layout
                : this;

        /* eslint-enable react/destructuring-assignment, react/prop-types */

        const {height} = props;
        return (
            <div
                style={{
                    height: `${height}px`,
                    marginBottom: `${height}px`,
                }}
            >
                <div id={this.elementId}>
                    <canvas
                        id={this.canvasId}
                        ref={ref => {
                            this.canvas = ref;
                        }}
                        style={{
                            pointerEvents: 'none',
                            position: 'absolute',
                            zIndex: '1',
                        }}
                    />
                </div>
            </div>
        );
    }
}

Map.defaultProps = {
    height: 800,
    layerNames: [],
};

Map.propTypes = {
    /**
     * The ID of this component, used to identify dash components
     * in callbacks. The ID needs to be unique across all of the
     * components in an app.
     */
    id: PropTypes.string.isRequired,
    /**
     * The data the Map component should render.
     * It should be JSON stringified before hand
     */
    data: PropTypes.string.isRequired,
    /**
     * The height of the Map component
     */
    height: PropTypes.number,
    /**
     * The name of individual layers
     */
    layerNames: PropTypes.arrayOf(PropTypes.string),
};

export default Map;
