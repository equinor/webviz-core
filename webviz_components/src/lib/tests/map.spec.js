import React from 'react';
import renderer from 'react-test-renderer';
import Map, {makeFlowLayers, make2DLayers} from '../components/Map';

const dataNoFlow = {
    i: {'0': 0, '1': 1, '2': 0, '3': 1},
    j: {'0': 0, '1': 0, '2': 1, '3': 1},
    k: {'0': 0, '1': 0, '2': 0, '3': 0},
    x0: {'0': 0, '1': 1, '2': 0, '3': 1},
    y0: {'0': 0, '1': 0, '2': 1, '3': 1},
    x1: {'0': 1, '1': 2, '2': 1, '3': 2},
    y1: {'0': 0, '1': 0, '2': 1, '3': 1},
    x2: {'0': 1, '1': 2, '2': 1, '3': 2},
    y2: {'0': 1, '1': 1, '2': 2, '3': 2},
    x3: {'0': 0, '1': 1, '2': 0, '3': 1},
    y3: {'0': 1, '1': 1, '2': 2, '3': 2},
    value: {'0': 1, '1': 0, '2': 4, '3': 2},
};

const dataWithFlow = {
    ...dataNoFlow,
    'FLOWI+': {'0': 0.005, '1': 0.002, '2': 0.001, '3': 0.004},
    'FLOWJ+': {'0': 0.0025, '1': 0.0045, '2': 0.0025, '3': 0.0035},
};

describe('Map', () => {
    it('should render Map', () => {
        const tree = renderer.create(
            <Map id="simple-map" data={JSON.stringify(dataWithFlow)} />
        );
        expect(tree.toJSON()).toMatchSnapshot();
    });
    it('should render Map as a dash component', () => {
        const dashTree = renderer.create(
            <Map
                id="simple-map"
                data=""
                _dashprivate_layout={{
                    props: {
                        id: 'simple-map',
                        data: JSON.stringify(dataWithFlow),
                    },
                }}
            />
        );
        const tree = renderer.create(
            <Map id="simple-map" data={JSON.stringify(dataWithFlow)} />
        );
        expect(tree.toString()).toEqual(dashTree.toString());
    });
});

describe('makeFlowLayers', () => {
    it('should match snapshot', () => {
        const layers = makeFlowLayers(dataWithFlow);
        expect(layers).toMatchSnapshot();
    });
});

describe('make2DLayers', () => {
    it('should match snapshot', () => {
        const layers = make2DLayers(dataNoFlow);
        expect(layers).toMatchSnapshot();
    });
});
