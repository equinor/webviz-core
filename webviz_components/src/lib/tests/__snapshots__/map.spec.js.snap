// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map should render Map 1`] = `
<div
  style={
    Object {
      "height": "800px",
      "marginBottom": "800px",
    }
  }
>
  <div
    id="container-simple-map"
  >
    <canvas
      id="canvas-simple-map"
      style={
        Object {
          "pointerEvents": "none",
          "position": "absolute",
          "zIndex": "1",
        }
      }
    />
  </div>
</div>
`;

exports[`make2DLayers should match snapshot 1`] = `
Array [
  Array [
    Object {
      "i": 0,
      "j": 0,
      "k": 0,
      "points": Array [
        Array [
          0,
          0,
        ],
        Array [
          1,
          0,
        ],
        Array [
          1,
          1,
        ],
        Array [
          0,
          1,
        ],
      ],
      "value": 1,
    },
    Object {
      "i": 1,
      "j": 0,
      "k": 0,
      "points": Array [
        Array [
          1,
          0,
        ],
        Array [
          2,
          0,
        ],
        Array [
          2,
          1,
        ],
        Array [
          1,
          1,
        ],
      ],
      "value": 0,
    },
    Object {
      "i": 0,
      "j": 1,
      "k": 0,
      "points": Array [
        Array [
          0,
          1,
        ],
        Array [
          1,
          1,
        ],
        Array [
          1,
          2,
        ],
        Array [
          0,
          2,
        ],
      ],
      "value": 4,
    },
    Object {
      "i": 1,
      "j": 1,
      "k": 0,
      "points": Array [
        Array [
          1,
          1,
        ],
        Array [
          2,
          1,
        ],
        Array [
          2,
          2,
        ],
        Array [
          1,
          2,
        ],
      ],
      "value": 2,
    },
  ],
]
`;

exports[`makeFlowLayers should match snapshot 1`] = `
Array [
  Array [
    Object {
      "FLOWI+": 0.005,
      "FLOWI-": 0,
      "FLOWJ+": 0.0025,
      "FLOWJ-": 0,
      "i": 0,
      "j": 0,
      "k": 0,
      "points": Array [
        Array [
          0,
          0,
        ],
        Array [
          1,
          0,
        ],
        Array [
          1,
          1,
        ],
        Array [
          0,
          1,
        ],
      ],
      "value": 1,
    },
    Object {
      "FLOWI+": 0.002,
      "FLOWI-": 0.005,
      "FLOWJ+": 0.0045,
      "FLOWJ-": 0,
      "i": 1,
      "j": 0,
      "k": 0,
      "points": Array [
        Array [
          1,
          0,
        ],
        Array [
          2,
          0,
        ],
        Array [
          2,
          1,
        ],
        Array [
          1,
          1,
        ],
      ],
      "value": 0,
    },
    Object {
      "FLOWI+": 0.001,
      "FLOWI-": 0,
      "FLOWJ+": 0.0025,
      "FLOWJ-": 0.0025,
      "i": 0,
      "j": 1,
      "k": 0,
      "points": Array [
        Array [
          0,
          1,
        ],
        Array [
          1,
          1,
        ],
        Array [
          1,
          2,
        ],
        Array [
          0,
          2,
        ],
      ],
      "value": 4,
    },
    Object {
      "FLOWI+": 0.004,
      "FLOWI-": 0.001,
      "FLOWJ+": 0.0035,
      "FLOWJ-": 0.0045,
      "i": 1,
      "j": 1,
      "k": 0,
      "points": Array [
        Array [
          1,
          1,
        ],
        Array [
          2,
          1,
        ],
        Array [
          2,
          2,
        ],
        Array [
          1,
          2,
        ],
      ],
      "value": 2,
    },
  ],
]
`;
